# Tarea - Programación lineal


### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables:

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

In [23]:
import numpy as np
import scipy.optimize as opt
c = np.array([0.013, 0.008, 0.010, 0.002, 0.005, 0.001])

A = np.array([
    [-0.10, -0.20, -0.15, -0.00, -0.04, -0.00],
    [-0.08, -0.10, -0.11, -0.01, -0.01,  0.00],
    [ 0.001, 0.005, 0.005, 0.10,  0.15,  0.00],
    [ 0.002, 0.005, 0.007, 0.002, 0.008, 0.00],
    [ 1.0,   1.0,   1.0,   1.0,   1.0,   1.0],
    [-1.0,  -1.0,  -1.0,  -1.0,  -1.0,  -1.0],
])
b = np.array([-8.0, -6.0, 2.0, 0.4, 100.0, -100.0])

opt.linprog(c, A, b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 0.52
              x: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00  0.000e+00
                   4.000e+01]
            nit: 2
          lower:  residual: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00
                              0.000e+00  4.000e+01]
                 marginals: [ 6.400e-03  0.000e+00  1.300e-03  3.000e-04
                              3.300e-03  0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 4.000e+00  0.000e+00  1.700e+00  1.000e-01
                              0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -7.000e


- Se obtiene una combinación mínima en costo de ingredientes para 100 g de comida.
- Solución: 60 g de carne + 40 g de gel.
- Costo mínimo = 0.52 u.m.
- Cumple proteína, grasa, fibra, sal y el total exacto de 100 g.

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente.


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla.

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

In [24]:
c = np.array([5,2,7, 3,6,6, 6,1,2])
A = np.array([
    [1,1,1, 0,0,0, 0,0,0],
    [0,0,0, 1,1,1, 0,0,0],
    [0,0,0, 0,0,0, 1,1,1],
    [-1,0,0, -1,0,0, -1,0,0],
    [0,-1,0, 0,-1,0, 0,-1,0],
    [0,0,-1, 0,0,-1, 0,0,-1],
])
b = np.array([80,40,60, -70,-40,-70])

opt.linprog(c, A, b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 540.0
              x: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01  0.000e+00
                   0.000e+00  0.000e+00  0.000e+00  6.000e+01]
            nit: 6
          lower:  residual: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01
                              0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+01]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+00  1.000e+00  6.000e+00  4.000e+00
                              0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf        inf        inf
                                    inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00  0.000e+00  0.0

- Cada planta distribuye su energía a las ciudades según capacidad y demanda.
- Solución:  
  - P1 = 30 GDL, 40 León, 10 Morelia  
  - P2 = 40 GDL  
  - P3 = 60 Morelia  
- Costo mínimo = 540 u.m.
- Todas las demandas quedan satisfechas y no se exceden las ofertas.

### Fábrica de focos

Una fábrica produce focos incandesentes y los vende a 4.5 u.m., además de focos ahorradores que vende a 6 u.m. cada uno. La producción está limitada a 400 focos incandesentes y 300 focos ahorradores al día, además de no poder producir más de 500 focos en total.

Si la fábrica vende toda su producción, determina cuántos focos incandesentes y cuántos focos ahorradores debe producir para obtener los máximos ingresos posibles y cuáles serían éstos.

In [25]:
c = np.array([-4.5, -6.0])
A = np.array([[1,0],[0,1],[1,1]])
b = np.array([400,300,500])

opt.linprog(c, A, b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -2700.0
              x: [ 2.000e+02  3.000e+02]
            nit: 1
          lower:  residual: [ 2.000e+02  3.000e+02]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 2.000e+02  0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -1.500e+00 -4.500e+00]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

- Se busca maximizar el ingreso por venta de focos incandescentes y ahorradores.
- Restricciones: 400 incandescentes, 300 ahorradores, máximo 500 en total.
- Solución: 200 incandescentes y 300 ahorradores.
- Ingreso máximo = 2700 u.m.

### Problema de compra al mayoreo
Un frutero necesita 16 cajas de naranjas, 5 de plátanos y 20 de manzanas. Dos mayoristas pueden suministrarle para satisfacer sus necesidades, pero sólo venden la fruta en contenedores completos. El mayorista A envía en cada contenedor 8 cajas de naranjas, 1 de plátanos y 2 de manzanas. El mayorista B envía en cada contenedor 2 cajas de naranjas, una de plátanos y 7 de manzanas. Sabiendo que cada contenedor del mayorista A tiene un costo de 150 u.m. y el contenedor del mayorista B de 300 u.m., calcular cuántos contenedores habrá de comprar a cada mayorista, con objeto de reducir al mínimo el costo de lo solicitado.

In [26]:
c = np.array([150,300])
A = np.array([[-8,-2],[-1,-1],[-2,-7]])
b = np.array([-16,-5,-20])

opt.linprog(c, A, b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 1050.0
              x: [ 3.000e+00  2.000e+00]
            nit: 3
          lower:  residual: [ 3.000e+00  2.000e+00]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 1.200e+01  0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -9.000e+01 -3.000e+01]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

- Se requieren cajas de frutas cumpliendo mínimo de naranjas, plátanos y manzanas.
- Cada mayorista vende contenedores completos con combinaciones distintas.
- Solución entera: 3 contenedores de A y 2 de B.
- Costo mínimo = 1050 u.m.
- Se cumplen todas las cantidades requeridas con el menor gasto.